src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/ManagedIndexSweeper.kt

```diff
@@ -0,0 +1,370 @@
+            initBackgroundSweep()
+        }
+        clusterService.clusterSettings.addSettingsUpdateConsumer(INDEX_STATE_MANAGEMENT_ENABLED) {
+            indexStateManagementEnabled = it
+            if (!indexStateManagementEnabled) disable() else enable()
+        }
+    }
+
+    override fun onMaster() {
+        // Init background sweep when promoted to being master
+        initBackgroundSweep()
+    }
+
+    override fun offMaster() {
+        // Cancel background sweep when demoted from being master
+        scheduledFullSweep?.cancel()
+    }
+
+    override fun executorName(): String {
+        return ThreadPool.Names.MANAGEMENT
+    }
+
+    override fun afterStart() {
+        initBackgroundSweep()
+    }
+
+    override fun beforeStop() {
+        scheduledFullSweep?.cancel()
+    }
+
+    private fun enable() {
+        initBackgroundSweep()
+        indexStateManagementEnabled = true
+    }
+
+    private fun disable() {
+        scheduledFullSweep?.cancel()
+        indexStateManagementEnabled = false
+    }
+
+    private fun isIndexStateManagementEnabled(): Boolean = indexStateManagementEnabled == true
+
+    /**
+     * Background sweep process that periodically sweeps for updates to ManagedIndices
+     *
+     * This background sweep will only be initialized if the local node is the elected master node.
+     * Creates a runnable that is executed as a coroutine in the shared pool of threads on JVM.
+     */
+    private fun initBackgroundSweep() {
+        // If ISM is disabled return early
+        // TODO: Should we have separate ISM-enabled and Sweeper-enabled flags?
+        if (!isIndexStateManagementEnabled()) return
+
+        // Do not setup background sweep if we're not the elected master node
+        if (!clusterService.state().nodes().isLocalNodeElectedMaster) return
+
+        // Cancel existing background sweep
+        scheduledFullSweep?.cancel()
+
+        // Setup an anti-entropy/self-healing background sweep, in case we fail to create a ManagedIndex job
+        val scheduledSweep = Runnable {
+            val elapsedTime = getFullSweepElapsedTime()
+
+            // Rate limit to at most one full sweep per sweep period
+            // The schedule runs may wake up a few milliseconds early
+            // Delta will be giving some buffer on the schedule to allow waking up slightly earlier
+            val delta = sweepPeriod.millis - elapsedTime.millis
+            if (delta < BUFFER) { // give 20ms buffer.
+                launch {
+                    try {
+                        logger.debug("Performing background sweep of managed indices")
+                        sweep()
+                    } catch (e: Exception) {
+                        logger.error("Failed to sweep managed indices", e)
+                    }
+                }
+            }
+        }
+
+        scheduledFullSweep = threadPool.scheduleWithFixedDelay(scheduledSweep, sweepPeriod, ThreadPool.Names.SAME)
```

Comment:
just want to confirm, even we schedule this runnable in scheduler threadpool, the sweep() function in launch{} block will be executed in another thread managed by Kotlin, right?

Reply:
Yes, it will use the ForkJoinPool.commonPool that exists by default.
@vinooamzn Can you confirm?

Reply:
Confirmed